metadata-ingestion/tests/unit/stateful_ingestion/provider/test_provider.py

```diff
@@ -0,0 +1,193 @@
+
+from avrogen.dict_wrapper import DictWrapper
+
+from datahub.emitter.mcp import MetadataChangeProposalWrapper
+from datahub.ingestion.api.common import PipelineContext
+from datahub.ingestion.api.ingestion_job_checkpointing_provider_base import (
+    CheckpointJobStateType,
+    IngestionCheckpointingProviderBase,
+    JobId,
+)
+from datahub.ingestion.source.state.checkpoint import Checkpoint
+from datahub.ingestion.source.state.sql_common_state import (
+    BaseSQLAlchemyCheckpointState,
+)
+from datahub.ingestion.source.state.usage_common_state import (
+    BaseTimeWindowCheckpointState,
+)
+from datahub.ingestion.source.state_provider.datahub_ingestion_checkpointing_provider import (
+    DatahubIngestionCheckpointingProvider,
+)
+from datahub.ingestion.source.state_provider.file_ingestion_checkpointing_provider import (
+    FileIngestionCheckpointingProvider,
+)
+from datahub.ingestion.source.state_provider.state_provider_registry import (
+    ingestion_checkpoint_provider_registry,
+)
+from tests.test_helpers.type_helpers import assert_not_null
+
+
+class TestDatahubIngestionCheckpointProvider(unittest.TestCase):
+    # Static members for the tests
+    pipeline_name: str = "test_pipeline"
+    job_names: List[JobId] = [JobId("job1"), JobId("job2")]
+    run_id: str = "test_run"
+
+    def setUp(self) -> None:
+        self._setup_mock_graph()
+        self._create_providers()
+
+    def _setup_mock_graph(self) -> None:
+        """
+        Setup monkey-patched graph client.
+        """
+        self.patcher = patch(
+            "datahub.ingestion.graph.client.DataHubGraph", autospec=True
+        )
+        self.addCleanup(self.patcher.stop)
+        self.mock_graph = self.patcher.start()
+        # Make server stateful ingestion capable
+        self.mock_graph.get_config.return_value = {"statefulIngestionCapable": True}
+        # Bind mock_graph's emit_mcp to testcase's monkey_patch_emit_mcp so that we can emulate emits.
+        self.mock_graph.emit_mcp = types.MethodType(
+            self.monkey_patch_emit_mcp, self.mock_graph
+        )
+        # Bind mock_graph's get_latest_timeseries_value to monkey_patch_get_latest_timeseries_value
+        self.mock_graph.get_latest_timeseries_value = types.MethodType(
+            self.monkey_patch_get_latest_timeseries_value, self.mock_graph
+        )
+        # Tracking for emitted mcps.
+        self.mcps_emitted: Dict[str, MetadataChangeProposalWrapper] = {}
+
+    def _create_providers(self) -> None:
+        ctx: PipelineContext = PipelineContext(
+            run_id=self.run_id, pipeline_name=self.pipeline_name
+        )
+        ctx.graph = self.mock_graph
+        self.providers: List[IngestionCheckpointingProviderBase] = []
+        for provider_name in ingestion_checkpoint_provider_registry.mapping.keys():
+            provider_cls = ingestion_checkpoint_provider_registry.get(provider_name)
+            if provider_cls == DatahubIngestionCheckpointingProvider:
```

Comment:
why are we iterating over `ingestion_checkpoint_provider_registry` here? we can just instantiate the two providers directly